kernel/safe_typing.ml

```diff
@@ -683,12 +683,14 @@ let build_module_body params restype senv =
     with one extra component and some updated fields
     (constraints, required, etc) *)
 
+let allow_delayed_constants = ref false
+
 let propagate_senv newdef newenv newresolver senv oldsenv =
   let now_cst, later_cst = List.partition Future.is_val senv.future_cst in
   (* This asserts that after Paral-ITP, standard vo compilation is behaving
    * exctly as before: the same universe constraints are added to modules *)
-  if !Flags.compilation_mode = Flags.BuildVo &&
-     !Flags.async_proofs_mode = Flags.APoff then assert(later_cst = []);
+  if not !allow_delayed_constants && later_cst <> [] then
+    CErrors.anomaly ~label:"safe_typing" Pp.(str "unresolved opaque futures in .vo mode");
```

Comment:
This message is not exact. The later_cst may be non-empty but all futures may have been jointed. If the list is empty it means that no proof has been delayed (i.e. things got executed in the very same order of Coq 8.4, and with the most "conservative" code paths).

I'd say "(proper) Future.t for opaque constants were created even if -async-proofs was off"

Reply:
Ok, thanks Enrico, updating the message.